### ISSUE TITLE
Missed optimization: Reorder deallocations before allocations of unrelated memory

### ISSUE THREAD

**Issue by davidstone** [llvm/llvm-project#68365]
Note: I am not certain which language standards allow this optimization, if any.

If we could move deallocations to occur before allocations of unrelated memory, it would reduce peak memory usage and potentially improve cache performance.

In other words, in the following example I would like to transform `bad` into `good`.

```cpp
int * bad(int * ptr) {
	auto temp = static_cast<int *>(operator new(10 * sizeof(int)));
	operator delete(ptr);
	return temp;
}

int * good(int * ptr) {
	operator delete(ptr);
	return static_cast<int *>(operator new(10 * sizeof(int)));
}
```

https://godbolt.org/z/WGq1jsdGd

With similar ability for things like `malloc` and `free`.

**Comment by Endilll** [comment]
I believe the primary rule is the following ([intro.execution-9](http://eel.is/c++draft/basic.exec#intro.execution-9)):
> Every value computation and side effect associated with a full-expression is sequenced before every value computation and side effect associated with the next full-expression to be evaluated.

[expr.new#13](http://eel.is/c++draft/expr.new#13) gives us some leeway, but not enough in my reading:
> An implementation is allowed to omit a call to a replaceable global allocation function ([[new.delete.single]](http://eel.is/c++draft/new.delete.single), [[new.delete.array]](http://eel.is/c++draft/new.delete.array))[.](http://eel.is/c++draft/expr.new#13.sentence-1) When it does so, the storage is instead provided by the implementation or provided by extending the allocation of another [new-expression](http://eel.is/c++draft/expr.new#nt:new-expression)[.](http://eel.is/c++draft/expr.new#13.sentence-2)

**Comment by Endilll** [comment]
CC @AaronBallman @shafik

**Comment by shafik** [comment]
I think this is allowed but I don't know what are the practical considerations.

pinging @zygoloid who was one of the authors of [Clarifying Memory Allocation](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3664.html).

**Comment by zygoloid** [comment]
The guidance from WG21 that the proposal wording was trying to follow was, as I recall:
- only call `operator new` at a point where a *new-expression* is evaluated
- only call `operator delete` at a point where a *delete-expression* is evaluated
- do not increase the high-water-mark for allocated memory
The first two points were trying to avoid memory allocations or deallocations being reordered into e.g.
a location where a mutex is held, or past some query into the allocator asking how much memory is in use. And the proposal wording is a stricter form of those rules.

So I don't think we can do this as a conforming optimization. We could do it if we could first turn one of the allocations into a `malloc`/`free` pair, because they're not replaceable and hence not considered observable.

**Comment by davidstone** [comment]
One of the primary use cases that I care about optimizing here are the allocation and deallocation functions of `std::allocator`. Given that the standard says it's unspecified when or how it calls `operator new` and `operator delete`, does that also have sufficient room to reorder them around each other?

**Comment by zygoloid** [comment]
Yes, we have extra freedom for `std::allocator` compared to `new` and `delete` operators. Currently `allocator` uses `__builtin_operator_new`/`delete`, which are specified as being optimizable like `new`/`delete` expressions. But I suppose we could change that and allow extra optimizations for the builtins.

**Comment by davidstone** [comment]
I'd ideally like a slightly more general optimization, which is that deallocation can be moved earlier than anything we know does not access the memory being deallocated (and have us 'know' that allocation does not do that for any arbitrary memory). It can be tricky in a few cases to decide when that optimization should be done, for instance in

```cpp
#include <cstdlib>

int x;
int y;

void a(void * ptr) {
	free(ptr);
	x = 1;
	y = 2;
}

void b(void * ptr) {
	x = 1;
	free(ptr);
	y = 2;
}

void c(void * ptr) {
	x = 1;
	y = 2;
	free(ptr);
}
```

https://godbolt.org/z/7EWr795Pj

my ordering preference is c, then a, then b, just because of the tail call optimization actually makes having it be the "last" conceptual thing done actually happen probably earlier than with any other ordering. But in theory it would be nice if all three of these implementations could be reordered into each other. However, there might be an implementation of global allocators that allows the user to modify a global variable to change their behavior thus making this optimization unsound.

**Comment by gbdngb12** [comment]
> Note: I am not certain which language standards allow this optimization, if any.
> 
> If we could move deallocations to occur before allocations of unrelated memory, it would reduce peak memory usage and potentially improve cache performance.
> 
> In other words, in the following example I would like to transform `bad` into `good`.
> 
> ```c++
> int * bad(int * ptr) {
> 	auto temp = static_cast<int *>(operator new(10 * sizeof(int)));
> 	operator delete(ptr);
> 	return temp;
> }
> 
> int * good(int * ptr) {
> 	operator delete(ptr);
> 	return static_cast<int *>(operator new(10 * sizeof(int)));
> }
> ```
> 
> https://godbolt.org/z/WGq1jsdGd
> 
> With similar ability for things like `malloc` and `free`.

Is there a compiler flag that enforces this specific ordering?
